src/drivers/rc_input/RCInput.cpp

```diff
@@ -0,0 +1,816 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "RCInput.hpp"
+
+#include "crsf_telemetry.h"
+
+using namespace time_literals;
+
+#if defined(SPEKTRUM_POWER)
+static bool bind_spektrum(int arg);
+#endif /* SPEKTRUM_POWER */
+
+work_s RCInput::_work = {};
+
+RCInput::RCInput(bool run_as_task) :
+	_cycle_perf(perf_alloc(PC_ELAPSED, "rc_input cycle time")),
+	_publish_interval_perf(perf_alloc(PC_INTERVAL, "rc_input publish interval"))
+{
+	// rc input, published to ORB
+	_rc_in.input_source = input_rc_s::RC_INPUT_SOURCE_PX4FMU_PPM;
+
+	// initialize it as RC lost
+	_rc_in.rc_lost = true;
+
+	// initialize raw_rc values and count
+	for (unsigned i = 0; i < input_rc_s::RC_INPUT_MAX_CHANNELS; i++) {
+		_raw_rc_values[i] = UINT16_MAX;
+	}
+}
+
+RCInput::~RCInput()
+{
+	orb_unadvertise(_to_input_rc);
+
+#ifdef RC_SERIAL_PORT
+	dsm_deinit();
+#endif
+
+	if (_crsf_telemetry) {
+		delete (_crsf_telemetry);
+	}
+
+	perf_free(_cycle_perf);
+	perf_free(_publish_interval_perf);
+}
+
+int
+RCInput::init()
+{
+	_adc_sub = orb_subscribe(ORB_ID(adc_report));
+
+#ifdef RC_SERIAL_PORT
+
+#  ifdef RF_RADIO_POWER_CONTROL
+	// power radio on
+	RF_RADIO_POWER_CONTROL(true);
+#  endif
+	_vehicle_cmd_sub = orb_subscribe(ORB_ID(vehicle_command));
+	// dsm_init sets some file static variables and returns a file descriptor
+	_rcs_fd = dsm_init(RC_SERIAL_PORT);
+	// assume SBUS input and immediately switch it to
+	// so that if Single wire mode on TX there will be only
+	// a short contention
+	sbus_config(_rcs_fd, board_supports_single_wire(RC_UXART_BASE));
+#  ifdef GPIO_PPM_IN
+	// disable CPPM input by mapping it away from the timer capture input
+	px4_arch_unconfiggpio(GPIO_PPM_IN);
+#  endif
+#endif
+
+	return 0;
+}
+
+int
+RCInput::task_spawn(int argc, char *argv[])
+{
+	bool run_as_task = false;
+	bool error_flag = false;
+
+	int myoptind = 1;
+	int ch;
+	const char *myoptarg = nullptr;
+
+	while ((ch = px4_getopt(argc, argv, "t", &myoptind, &myoptarg)) != EOF) {
+		switch (ch) {
+		case 't':
+			run_as_task = true;
+			break;
+
+		case '?':
+			error_flag = true;
+			break;
+
+		default:
+			PX4_WARN("unrecognized flag");
+			error_flag = true;
+			break;
+		}
+	}
+
+	if (error_flag) {
+		return -1;
+	}
+
+
+	if (!run_as_task) {
+
+		/* schedule a cycle to start things */
+		int ret = work_queue(HPWORK, &_work, (worker_t)&RCInput::cycle_trampoline, nullptr, 0);
+
+		if (ret < 0) {
+			return ret;
+		}
+
+		_task_id = task_id_is_work_queue;
+
+	} else {
+
+		/* start the IO interface task */
+
+		_task_id = px4_task_spawn_cmd("rc_input",
+					      SCHED_DEFAULT,
+					      SCHED_PRIORITY_SLOW_DRIVER,
+					      1000,
+					      (px4_main_t)&run_trampoline,
+					      nullptr);
+
+		if (_task_id < 0) {
+			_task_id = -1;
+			return -errno;
+		}
+	}
+
+	// wait until task is up & running (the mode_* commands depend on it)
```

Comment:
left-over comment. I don't think we need to wait here.

Reply:
Fixed

Reply:
Dropped